### ISSUE TITLE
set_interrupt loses IRQ edges under load — observable stalls in guest I/O

### ISSUE THREAD

**Issue by olofj** [olofj/bhx#195]
### Symptom

Long stall in guest I/O — observed during a manual openSUSE install on V2.1:

- Installer froze mid-package at \"55% installed, 45 packages remaining\" for 12+ minutes.
- Daemon's \`bhx_notify_events_total\` and \`bhx_dispatch_passes_total\` were flatlined across multiple samples (zero guest QUEUE_NOTIFY traffic).
- Daemon log dispatcher silence for the entire window (no STATUS transitions, no new queue setups).
- Guest console responsive to navigation (could see progress bar, hit buttons).
- A few keystrokes through the chip-UART unblocked the install: \`bhx_notify_events_total\` immediately jumped from 174573 to 219119 (+45k in 2 s), install completed normally.

The pattern — host I/O completion silently dropped, unrelated IRQ later nudges the kernel into re-checking — is the textbook signature of a missed completion interrupt.

### Root cause

\`src/virtio/interrupt.rs::set_interrupt\`:

\`\`\`rust
// BUG PRESERVED: sets only our interrupt, doesn't OR with existing
ptr::write_volatile(reg, 1u32 << (interrupt_number - 5));
atomic::fence(Ordering::SeqCst);
ptr::write_volatile(reg, 0u32);
\`\`\`

PLIC pending bit is asserted then de-asserted within a sub-microsecond window. If the X280 hart is mid-claim of another interrupt or hasn't scanned the pending register in this cycle, the edge is lost. Additionally, the \`write 0\` clobbers ANY sibling-device pending bits in the same PLIC register (4+ virtio devices per L2CPU share one register).

### History — why naive fixes don't work

Inherited from the C++ original at \`tt-bh-linux/console/virtiodevice.hpp::set_interrupt\`. The C++ history records three iterations of this code, each addressing a real failure mode:

| Commit | Date | What changed | Why |
|--------|------|--------------|-----|
| \`5aa559d\` | 2025-08-02 | Comment out PLIC clear in \`ack_interrupt\` | Linux PLIC driver patches were clearing the bit kernel-side |
| \`bc6a858\` | 2025-08-05 | Per-descriptor IRQs (not per-batch) | Block device \"appears hung\" with batched IRQs — same symptom this issue reports |
| \`e024439\` | 2025-09-27 | Fully NO-OP \`ack_interrupt\` | Clearing \`interrupt_status\` from daemon was \"too fast\" — wiped next IRQ's VRING bit before kernel handler ran |
| \`93ea911\` | 2025-11-27 | Add \`write 0\` to PLIC pending immediately in \`set_interrupt\` | \"could let us get rid of the PLIC kernel changes… Previously we were doing it on the device side, but I believe it was happening too late, and that was causing an interrupts storm on the PLIC\" |

**Critical inference from \`93ea911\`**: the X280 PLIC's pending bit does NOT auto-clear on hart claim. Without an explicit daemon clear, the kernel re-enters IRQ context in a tight loop. The C++ team had three viable architectures and picked the least-bad:

1. **Kernel-side clear** (Linux PLIC driver patches) — works but requires carrying kernel patches.
2. **Daemon-side clear in \`ack_interrupt\`** — racy in both directions: too-early wipes next IRQ's status, too-late causes storm.
3. **Daemon-side clear immediately in \`set_interrupt\`** — current code. Avoids storm. Loses edges under load.

(3) was the implicit accept-the-rare-missed-IRQ trade. Our install stall proves \"rare\" is unacceptable under heavy block load.

### Why this surfaces on V2.1 but wasn't loud on V1

V1's chronic kick-ring-overflow noise (#184) caused enough other I/O failures that this latent bug couldn't be cleanly attributed. V2.1 cleaned up the dispatch path; the missed-IRQ class is now the tallest pole and visible.

### Naive fix that would regress

\`\`\`rust
// THIS IS WRONG — would re-introduce the interrupts storm 93ea911 fixed.
let cur = ptr::read_volatile(reg);
ptr::write_volatile(reg, cur | (1u32 << (interrupt_number - 5)));
// no write 0
\`\`\`

OR-in with no clear: PLIC pending bit stays asserted forever (no auto-clear). Kernel claims, returns from IRQ, immediately re-IRQs, tight loop until daemon next pokes (which only updates the same bit). Confirmed bad by C++ history.

### Fix candidates that respect the constraints

All three need to **keep the PLIC bit asserted long enough for the X280 to latch the edge AND clear it before the kernel storms**. The C++ team's three architectures map onto the design space:

**A. Delayed clear via short busy-wait in \`set_interrupt\`.** Set bit, fence, busy-wait ~5-20 μs (tunable), write 0. Widens the latch window without exposing the daemon's ack-polling lag. Trivial to implement; cost is added IRQ-delivery latency. Tunable bound: should be longer than PLIC scan period, shorter than kernel IRQ-handler completion (~μs).

**B. Per-L2CPU dedicated ack-watcher thread.** One thread per L2CPU spinning on \`interrupt_ack\` MMIO. On transition to 1, clears PLIC pending bit and clears \`interrupt_ack\`. Couples clear timing to actual kernel handler completion (not to a guess). Cost: 4 extra threads per daemon, one cache line spinning per thread. The \`5aa559d\`/\`e024439\` C++ commits document why polling \`interrupt_ack\` from the existing worker poll loop was too slow — a dedicated thread polling tightly should not have that problem.

**C. Carry the PLIC-clear kernel patch.** Restore the kernel-side PLIC clear that the C++ team had before \`5aa559d\`. The simplest from a daemon-code POV but pushes the complexity into the kernel build. Doesn't work on stock distro kernels (a hard constraint per the V2.2 dead-end discussion in #191).

**Recommended**: (A) with a 10-20 μs busy-wait first, since it's the smallest possible diff and the testcase from this issue's gate will tell us empirically if the window is wide enough. If (A) still drops edges under sustained load, fall back to (B). (C) is ruled out by the stock-kernel constraint.

### Verifying the X280 PLIC's actual semantics

Before settling on a fix, confirm the auto-clear question with a microbenchmark: write \`1<<bit\` to pending, claim from hart, re-read pending. If pending returns to 0 without further writes, PLIC auto-clears on claim (and the C++ history's \"interrupts storm\" was misdiagnosed). If pending stays 1, we need the daemon-side clear path. Should be one debug subcommand, < 50 lines.

### Detection / regression gate

\`scripts/soak_irq_parity.py\` (filed in follow-up issue): counts daemon-side \`bhx_blk_interrupts_total\` vs guest-side \`/proc/interrupts\` virtio counts under a multi-job concurrent fio load with rng + 2× blk + net + console active. Asserts the gap stays within in-flight tolerance. Current code should fail; fixed code should pass.

### Severity

High — direct user-visible I/O stalls under sustained block load, with no automatic recovery (the install only resumed because a human happened to type into the console). The data isn't corrupted but workflows hang for indefinite wall time until something unrelated pokes the hart.

### Discovery context

Surfaced during a manual openSUSE install on \`virtio-v2\` (V2.1) — the install that motivated the V2.1 hammer campaign as a final shakedown. Pre-existing bug carried from the C++ rewrite (\`8b7d10e\`).

**Comment by olofj** [comment]
### Reproducer landed + first run confirms the bug

\`scripts/soak_irq_parity.py\` (commit 6e7c3fe) — runs a 16-job direct=1 fio randwrite workload inside the guest with rng + blk + net + console all active (max PLIC contention) and compares host-side IRQ fires vs guest-side IRQ receives every second.

**First run on current \`virtio-v2\` HEAD, 60 s duration:**

\`\`\`
=== summary ===
  duration                       : 60s
  samples                        : 60
  dispatch_passes_total          : 42530
  host_irqs_total (sum 4 kinds)  : 42533
  guest_irqs_total (virtio rows) : 42167
  cumulative gap                 : 366
  worst stall ratio (over 5s)    : 0.180 at t=16.0s

FAIL: missed-IRQ stall ratio 0.180 > tolerance 0.050 (worst at t=16.0s).
This is #195 — set_interrupt's de-assert window is too tight for the X280 PLIC
to latch under load.
\`\`\`

**Step-function loss episode at t=15 s:**

| time | host_irqs | guest_irqs | host_delta | guest_delta | gap |
|------|-----------|------------|------------|-------------|-----|
| t-1s | 9 960 | 10 063 | 720 | 718 | 0 |
| **t=15s** | **10 600** | **10 063** | **640** | **0** | **537** |
| t+1s | 11 304 | 10 781 | 704 | 718 | 523 |
| ... | ... | ... | ... | ... | (slow recovery) |
| t+18s | 17 696 | 17 201 | 708 | 714 | 495 |
| t+34s | 24 080 | 23 621 | 706 | 714 | 459 |
| end   | 42 533 | 42 167 | — | — | **366** |

That's **640 IRQs fired in one second, ZERO received by the kernel**, then ~3 minutes of slow drip-recovery as unrelated IRQs nudge the kernel to re-check ring state. Exactly the pattern observed in the original install-stall report.

Cumulative loss = 0.86 % across the full 60 s, but the burst-level loss rate (18 % in the worst 5 s window) is what actually matters — it's where users see actual stalls.

### Notes for the fix work

1. The CSV captures full per-second data so the test-bed is ready for any candidate fix — just rerun and compare. The smoke catches the bug in 60 s; production gate should be \`DURATION_SEC=300\`.
2. **Cumulative gap recovers partially over time** because the guest's block layer eventually times out and re-scans the ring (or a sibling IRQ nudges it). So a fix that takes the worst-case 5 s stall ratio to 0 will probably also drive the cumulative gap to single digits.
3. **Recommended X280 PLIC microbenchmark next** — write \`1<<bit\` to pending, claim from hart, re-read pending. If the bit clears without further writes, the PLIC auto-clears on claim and the C++ history's \`93ea911\` "interrupts storm" was misdiagnosed (and the simple OR-in fix is actually correct). If the bit stays at 1, we need delayed-clear or per-L2CPU ack-watcher.

**Comment by olofj** [comment]
### Resolution: 10 µs latch-window default (fixed in 100c6c4)

The latch-window sweep landed on a sharp phase transition at 9 µs. Picked 10 µs (one safety µs above the transition) as the production default.

#### Sweep data

| us | host_irqs | guest_irqs | guest/host | worst 5s stall |
|---|---|---|---|---|
| 0 | 42 580 | 42 241 | 0.992 | 17.9 % (loss) |
| 1-4 | ~42 400 | ~42 300 | ~0.99 | ~18 % (loss) |
| 5-8 | ~42 350 | ~42 700 | 1.00-1.03 | 17-19 % (mixed) |
| **9** | **42 257** | **53 218** | **1.26** | **0 % ✓** |
| 10 | 42 244 | 91 583 | 2.17 | 0 % |
| 15 | 42 089 | 304 468 | 7.23 | 0 % |
| 25 | 41 808 | 741 898 | 17.75 | 0 % |
| 50 | 41 170 | 1 815 964 | 44.1 | 0 % |
| 100 | 39 669 | 3 975 202 | 100.2 | 0 % |

#### Two facts the data settles

1. **The X280 PLIC re-samples the held source register continuously.** Wider latch → kernel-handler-per-fire ratio grows roughly linearly (100 µs → 100× storm). This is the \"interrupts storm\" the C++ commit \`93ea911\` was avoiding — they just chose a 0 µs window which sits on the wrong side of the phase transition.

2. **The phase boundary between loss and storm is sharp at 9 µs.** No clean \"1.0× ratio\" point exists — the system goes directly from 1.03× (some loss in windows balanced by surplus) to 1.26× (no loss, mild surplus). 10 µs is one safety µs above the transition; gives 2.17× kernel-handler amplification in exchange for zero misses. Each amplified invocation is cheap (kernel reads interrupt_status, sees VRING bit, finds nothing new on the redundant pass, returns).

#### What I tried that didn't work

- **Wait-for-claim** (poll the PLIC pending register from the daemon, break when it auto-clears, then de-assert) — implemented and reverted. Doesn't work because \`read_volatile(reg)\` on our source register returns what we just wrote, not the PLIC's derived pending state. The two are independent. Dispatcher blocked on the 1 ms safety timeout, throughput cratered to 12 IRQs/sec.
- **Simple OR-in** (don't clear at all) — ruled out by the C++ history, confirmed by the sweep: storm rate grows unboundedly with held duration.

#### Validation

- 60 s smoke at 10 µs default: PASS, zero stalls, 1.26× surplus.
- 5-min sustained at 9 µs: 211 667 fires / 266 663 receives, zero stalls.
- 2-min post-fix run (default 10 µs): 84 457 fires / 185 392 receives, zero stalls, cumulative gap 0.

#### Knob preserved

\`BHX_PLIC_LATCH_US=N\` env override stays for diagnostics. Setting it to 0 reproduces the pre-#195 bug, which is useful for re-validating \`scripts/soak_irq_parity.py\` going forward.

#### Follow-up: cost characterization

The 2.17× kernel-handler amplification means the guest now spends more cycles in IRQ context. Worth measuring: re-run \`scripts/bench/run_all.sh\` with the fix in place to confirm there's no measurable hit on the disk/net/console perf baselines I captured earlier in the V2.1 hammer campaign. If those stay within the 6-82% gains I measured pre-fix (V1 baseline), we're golden — the amplification cost is below the noise floor.
